objects/src/accounts/storage.rs

```diff
@@ -1,47 +1,83 @@
-use super::{Digest, Hasher, Word, ZERO};
+use crypto::merkle::{MerkleStore, NodeIndex, SimpleSmt};
+
+use super::{AccountError, Word};
 
 // TYPE ALIASES
 // ================================================================================================
 
-pub type StorageItem = (Word, Word);
+pub type StorageItem = (u8, Word);
 
 // ACCOUNT STORAGE
 // ================================================================================================
 
-/// Account storage is a key-value map where both keys and values are words (4 elements).
-///
-/// Internally, account storage uses a compact Sparse Merkle tree to store the data. To ensure that
-/// the data is randomly distributed across the tree, the index of item in the tree is derived by
-/// hashing the key.
-#[derive(Debug, Clone, Default, PartialEq, Eq)]
+/// Account storage is composed of two components. The first component is a simple sparse Merkle
+/// tree of depth 8 which is index addressable. This provides the user with 256 Word slots. Users
+/// that require additional storage can use the second component which is a `MerkleStore`. This
+/// will allow the user to store any Merkle structures they need.  This is achieved by storing the
+/// root of the Merkle structure as a leaf in the simple sparse merkle tree. When `AccountStorage`
+/// is serialized it will check to see if any of the leafs in the simple sparse Merkle tree are
+/// Merkle roots of other Merkle structures.  If any Merkle roots are found then the Merkle
+/// structures will be persisted in the `AccountStorage` `MerkleStore`.
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct AccountStorage {
-    // TODO: add compact SMT as backing storage
+    slots: SimpleSmt,
+    store: MerkleStore,
 }
 
 impl AccountStorage {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// Depth of the storage tree.
+    pub const STORAGE_TREE_DEPTH: u8 = 8;
+
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
     /// Returns a new instance of account storage initialized with the provided items.
-    pub fn new(_items: &[StorageItem]) -> Self {
-        Self {}
+    pub fn new(
+        items: Vec<StorageItem>,
+        store: MerkleStore,
+    ) -> Result<AccountStorage, AccountError> {
+        let slots = SimpleSmt::new(Self::STORAGE_TREE_DEPTH)
+            .expect("depth is valid")
+            .with_leaves(items.into_iter().map(|x| (x.0 as u64, x.1)))
+            .expect("`StorageItem` index is u8 - index within range");
+
+        Ok(Self { slots, store })
```

Comment:
The way this currently is, doesn't seem like this would ever return an error. One thing I wonder though: should we check that the provided list of storage items does not contain any duplicates? Duplicates in the list would most likely be a result of some other error which happened in the caller.

Reply:
That could make sense.  Is there a reason we don't check for duplicates in the `with_leaves(..)` function of the `SimpleSmt`?

Reply:
Hmm - no reason. It might be good to check for it there instead of here - but I'm fine either way.

Reply:
How are we defining duplicates?  Are we defining a duplicate as any two `StorageItem`'s which share the same storage index regardless of the storage value?

Reply:
Yep, I think any two items with the same key are duplicates.

Reply:
updated